presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java

```diff
@@ -75,30 +90,122 @@ public MemoryRevokingScheduler(
         this.taskManagementExecutor = taskManagementExecutor;
         this.memoryRevokingThreshold = memoryRevokingThreshold;
         this.memoryRevokingTarget = memoryRevokingTarget;
+        this.potentiallyOverflowingPools = synchronizedSet(new HashSet<>());
     }
 
     @PostConstruct
     public void start()
     {
+        registerPeriodicCheck();
+        registerPoolListeners();
+    }
+
+    @VisibleForTesting
+    void registerPeriodicCheck()
+    {
         taskManagementExecutor.scheduleWithFixedDelay(() -> {
             try {
-                requestMemoryRevokingIfNeeded(currentTasksSupplier.get());
+                requestMemoryRevokingIfNeeded(true);
             }
             catch (Throwable e) {
                 log.warn(e, "Error requesting system memory revoking");
             }
         }, 0, 1, TimeUnit.SECONDS);
     }
 
-    public void requestMemoryRevokingIfNeeded(Collection<SqlTask> sqlTasks)
+    @VisibleForTesting
+    void registerPoolListeners()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool)
+    {
+        try {
+            if (!memoryRevokingNeeded(memoryPool, memoryPool.getFreeBytes())) {
+                return;
+            }
+
+            if (potentiallyOverflowingPools.add(memoryPool)) {
+                log.debug("Scheduling check for %s", memoryPool);
+
+                taskManagementExecutor.execute(() -> {
+                    try {
+                        requestMemoryRevokingIfNeeded(false);
```

Comment:
There's a race here. The pool has been over-allocated, but the OperatorContext that called the pool potentially did not update it's revokable memory counter yet. If the `requestMemoryRevokingIfNeeded` is run too early, it might still revoke some memory (when there are multiple tasks with revokable memory or when the requesting one already had some revokable memory). Or, `requestMemoryRevokingIfNeeded` might see no revokable memory and will be void.